File: selected_coding_interview/codes/python/lc_415_add_strings.py
"""
File: lc_415_add_strings.py
Created Time: 2025-12-30
Author: krahets
"""

from include import *


# ===== Solution Code =====
class Solution:
    def addStrings(self, num1: str, num2: str) -> str:
        res = ""
        i, j, carry = len(num1) - 1, len(num2) - 1, 0
        while i >= 0 or j >= 0:
            n1 = int(num1[i]) if i >= 0 else 0
            n2 = int(num2[j]) if j >= 0 else 0
            tmp = n1 + n2 + carry
            carry = tmp // 10
            res = str(tmp % 10) + res
            i, j = i - 1, j - 1
        return "1" + res if carry else res


# ======= Test Case =======
test_input_num1 = "11"
test_input_num2 = "123"
expected_output = "134"

# ====== Driver Code ======
slt = Solution()
result = slt.addStrings(test_input_num1, test_input_num2)
print(result)

File: selected_coding_interview/codes/java/lc_415_add_strings/lc_415_add_strings.java
/*
* File: lc_415_add_strings.java
* Created Time: 2025-12-30
* Author: krahets
*/

package lc_415_add_strings;

import include.*;
import java.util.*;

// ===== Solution Code =====
    class Solution {
        public String addStrings(String num1, String num2) {
            StringBuilder res = new StringBuilder("");
            int i = num1.length() - 1, j = num2.length() - 1, carry = 0;
            while(i >= 0 || j >= 0){
                int n1 = i >= 0 ? num1.charAt(i) - '0' : 0;
                int n2 = j >= 0 ? num2.charAt(j) - '0' : 0;
                int tmp = n1 + n2 + carry;
                carry = tmp / 10;
                res.append(tmp % 10);
                i--; j--;
            }
            if(carry == 1) res.append(1);
            return res.reverse().toString();
        }
    }

public class lc_415_add_strings {
    public static void main(String[] args) {
        // ======= Test Case =======
        String test_input_num1 = "11";
        String test_input_num2 = "123";
        String expected_output = "134";

        // ====== Driver Code ======
        Solution slt = new Solution();
        var result = slt.addStrings(test_input_num1, test_input_num2);
        System.out.println(result);

    }
}
